src/main.rs

```diff
@@ -1,3 +1,67 @@
-fn main() {
-    println!("Hello, world!");
+use std::io;
+use std::io::{stdout, Write};
+use std::sync::Arc;
+
+use clap::Parser;
+use ethers::providers::{Http, Middleware, ProviderError, StreamExt};
+use ethers::signers::LocalWallet;
+use ethers::types::{Block, Transaction};
+
+use crate::cli::{ClientParseError, Opts};
+use crate::client::Client;
+
+pub mod cli;
+pub mod client;
+
+/// Represents global errors emitted from the bot process itself
+#[derive(Debug)]
+pub enum Error {
+    /// Error parsing client from provided arguments
+    ClientOptsError(ClientParseError),
+    /// Error in blockchain interaction
+    ProviderError(ProviderError),
+    /// Error in I/O
+    IOError(io::Error),
+}
+
+impl From<ClientParseError> for Error {
+    fn from(value: ClientParseError) -> Self {
+        Self::ClientOptsError(value)
+    }
+}
+
+impl From<ProviderError> for Error {
+    fn from(value: ProviderError) -> Self {
+        Self::ProviderError(value)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(value: io::Error) -> Self {
+        Self::IOError(value)
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    let opts: Opts = Opts::parse();
+    let client: Arc<Client<LocalWallet, Http>> = Arc::new(opts.try_into()?);
```

Comment:
I think this will need to change when websocket rpc support is added but that's probs a different pr

Reply:
Yeah I considered handling this for Websockets ahead of time but I think this allows more rapid local testing for now. Ideally we'd match on the provided RPC url and defer to the relevant `TryFrom` implementation in the `cli` module.

I'll add a comment to this effect.

Reply:
Consult ec9626d.